### ISSUE TITLE
Feature request: enable callback registration for C-based tools

### ISSUE THREAD

**Issue by wrwilliams** [kokkos/kokkos#3604]
C++ tools have the option to use the `Tools::Experimental::add_foo_callback` style of functions to enable/disable their individual callbacks. This means, among other things, that if a C++ tool wishes to link itself into the application or run via pure `LD_PRELOAD`, it can in principle choose to enable its callbacks whenever it sees fit. C tools, however, must use the `KOKKOS_PROFILE_LIBRARY` variable to register and may not individually enable/disable their callbacks.

What we would like to do in Score-P is the following:

* Link in our Kokkos event library (contains the actual `kokkosp_*` implementations) along with all other instrumentation libraries the user requests
* Enable some set of Kokkos profiling functions via Score-P environment variables (e.g. parallel regions, allocations, deep copies...)

What we currently must do:

* Link in our Kokkos management library only (contains the Kokkos instrumentation support code that's not called from `kokkosp_*` functions)
* Set the `KOKKOS_PROFILE_LIBRARY` variable to our event library
* Enable features via environment variable
* Have every `kokkosp_*` function check whether it's enabled and early-exit if it's not

I talked with @DavidPoliakoff last night about this, and whether having Kokkos simply check via `dlsym` for linked-in profiling functions would be sufficient, but I think that if there's a performance gain to be had from only registering the functions the user wants, that might be worth the added complexity of letting C-based tools register callbacks manually.

**Comment by jrmadsen** [comment]
So this kind of touches on a small personal peeve, C++ codes _routinely_ make accommodations for C codes but C codes _never_ appear willing to make accommodations for C++ codes. What exactly is preventing Score-P from writing a C++ header (and library if need be) that Kokkos users can use which makes the translation you need to call your C code?

**Comment by DavidPoliakoff** [comment]
I'm not opposed to this existing. If N tools all need to write the translation header, and N need to update it when we do something goofy, as opposed to us maintaining one...

Marginal, but I'll swing at this for 3.4

**Comment by jrmadsen** [comment]
I agree that if there is a demand, we should supply it. But adding new features based on a single "client" sometimes lends itself towards designing that new feature to fit the specifications of _that client only_ and things have to be re-designed again to accommodate the additional APIs. Not saying this will necessarily be the case here, just saying that having at least one more code that needs this feature might be helpful.

**Comment by wrwilliams** [comment]
> So this kind of touches on a small personal peeve, C++ codes _routinely_ make accommodations for C codes but C codes _never_ appear willing to make accommodations for C++ codes. What exactly is preventing Score-P from writing a C++ header (and library if need be) that Kokkos users can use which makes the translation you need to call your C code?

Honestly, the same thing that stops us from implementing Score-P in C++ as a whole: we would need to ensure that the library in question is (re)built as part of the executable using the same toolchain in order to actually be safe. Now, you could argue that it shouldn't matter for our purposes here, because we "just" need name mangling to call a C-style function inside a couple of namespaces. And you wouldn't be wrong--but if Kokkos intends to provide a C tools interface, 
I do think it makes sense to provide a more-or-less complete one, and especially when the main obstacle is de-namespacing perfectly good C functions.

With respect to the more general gripe, I think that once a C code goes to the effort of making accommodations for C++, it becomes a C++ code very quickly...

I think two things are true here:

* The C interface request for registration of callbacks is basically orthogonal to our intended usage of "install tool once and only once at link time": being able to register the callbacks without a shim to get us into the appropriate namespaces *enables* something that *should already work* but may not be well tested.
* I'm far more concerned about the implications of "install tool at link time not runtime" than about whether the registration methods are accessible from the C interface. If it comes down to it, I could write a quick PR that expands the C interface; the semantics of using the registration functions the way we want is a deeper question.

So yes, I would refocus things here a bit:

* Does the callback registration interface actually support a link-time tool registering its own callbacks without `KOKKOS_PROFILE_LIBRARY` or dlsym? Or is that something that looks reasonable but actually isn't?
* What are the preconditions for registering/deregistering a callback? Is post-`kokkos_init` necessary/sufficient?
* Are there other tools interested in controlling which callbacks they register by e.g environment variable? Is the performance win of not registering a callback meaningful compared to a callback that compares and returns?
* Are there other tools interested in link-time registration as an option? I initially assumed Score-P was the odd man out here because most other Kokkos tools are doing much less general instrumentation, but then there's the nSight plugins and other things that could reasonably be living in a link-time/`LD_PRELOAD` ecosystem.

If you guys know of other tool developers who'd be interested in this discussion please tag them.
